File: singleton.py
"""
Singleton Design Pattern Implementation (Python)

Multiple approaches to implement Singleton in Python
"""

# Approach 1: Using a metaclass
class SingletonMeta(type):
    """
    Metaclass for Singleton pattern
    """
    _instances = {}
    
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            instance = super().__call__(*args, **kwargs)
            cls._instances[cls] = instance
        return cls._instances[cls]


class Singleton(metaclass=SingletonMeta):
    """
    Singleton class using metaclass approach
    """
    
    def __init__(self):
        print("Singleton instance created!")
        self.value = None
    
    def do_something(self):
        print("Singleton is doing something!")


# Approach 2: Using a decorator
def singleton(cls):
    """
    Decorator to make a class a singleton
    """
    instances = {}
    
    def get_instance(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]
    
    return get_instance


@singleton
class SingletonDecorator:
    """
    Singleton class using decorator approach
    """
    
    def __init__(self):
        print("SingletonDecorator instance created!")
        self.value = None
    
    def do_something(self):
        print("SingletonDecorator is doing something!")


# Approach 3: Using __new__ method
class SingletonNew:
    """
    Singleton class using __new__ method
    """
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            print("SingletonNew instance created!")
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def do_something(self):
        print("SingletonNew is doing something!")


if __name__ == "__main__":
    # Test Approach 1
    print("=== Testing Metaclass Approach ===")
    s1 = Singleton()
    s2 = Singleton()
    print(f"Are both instances the same? {s1 is s2}")
    s1.do_something()
    
    print("\n=== Testing Decorator Approach ===")
    sd1 = SingletonDecorator()
    sd2 = SingletonDecorator()
    print(f"Are both instances the same? {sd1 is sd2}")
    sd1.do_something()
    
    print("\n=== Testing __new__ Approach ===")
    sn1 = SingletonNew()
    sn2 = SingletonNew()
    print(f"Are both instances the same? {sn1 is sn2}")
    sn1.do_something()

File: singleton.js
/**
 * Singleton Design Pattern Implementation (JavaScript)
 * 
 * Multiple approaches to implement Singleton in JavaScript
 */

// Approach 1: Using ES6 Class
class Singleton {
    constructor() {
        if (Singleton.instance) {
            return Singleton.instance;
        }
        
        console.log("Singleton instance created!");
        this.timestamp = new Date();
        Singleton.instance = this;
    }
    
    doSomething() {
        console.log("Singleton is doing something!");
    }
    
    static getInstance() {
        if (!Singleton.instance) {
            Singleton.instance = new Singleton();
        }
        return Singleton.instance;
    }
}

// Approach 2: Using Module Pattern (IIFE)
const SingletonModule = (function() {
    let instance;
    
    function createInstance() {
        console.log("SingletonModule instance created!");
        return {
            timestamp: new Date(),
            doSomething: function() {
                console.log("SingletonModule is doing something!");
            }
        };
    }
    
    return {
        getInstance: function() {
            if (!instance) {
                instance = createInstance();
            }
            return instance;
        }
    };
})();

// Approach 3: Using Object Literal (Simplest)
const SingletonObject = {
    timestamp: new Date(),
    doSomething: function() {
        console.log("SingletonObject is doing something!");
    }
};

// Example usage
if (typeof module !== 'undefined' && module.exports) {
    // Node.js environment
    module.exports = { Singleton, SingletonModule, SingletonObject };
    
    // Test the implementations
    console.log("=== Testing ES6 Class Approach ===");
    const s1 = new Singleton();
    const s2 = new Singleton();
    console.log("Are both instances the same?", s1 === s2);
    s1.doSomething();
    
    console.log("\n=== Testing Module Pattern ===");
    const sm1 = SingletonModule.getInstance();
    const sm2 = SingletonModule.getInstance();
    console.log("Are both instances the same?", sm1 === sm2);
    sm1.doSomething();
    
    console.log("\n=== Testing Object Literal ===");
    SingletonObject.doSomething();
}
